cpp/src/arrow/status.h

```diff
@@ -92,6 +92,7 @@ enum class StatusCode : char {
   SerializationError = 11,
   PythonError = 12,
   RError = 13,
+  SubsystemError = 14,
```

Comment:
I don't think that's the right approach. If I have a Python `ValueError`, I want it both to map to `StatusCode::Invalid` _and_ to retain the original Python exception.

Reply:
Yep, I tend to agree.  So two issues to improve on:
*  top-level codes should be orthogonal from subsytem codes
+  Potentially need the ability to directly link to underlying exception (e.g. PyObject for a python error)

Any other high level concepts/functionallty missing?

Reply:
Right. I also think subsystem codes can be optional (for example, Python would just use the underlying exception instead of defining integer codes).

So instead of having a subsystem code, we could have an opaque optional detail instance, e.g.:
```cpp
class StatusDetail {
 public:
  virtual ~StatusDetail();
  virtual std::string ToString() const = 0;
};
```

Reply:
Yeah, I came to the same conclusion after stepping away from the computer. Will try to update tonight.

Reply:
I've updated to just have a status detail, that wraps has a message accessor and unify the current implementation with a private implementation of the interface.  I can see arguments for keeping a separate message member as well, so I'm happy to add it back.

Reply:
I would be in favour of having a separate message member, so that the detail can remain null if there's no error-specific detail to carry.

Reply:
Added msg field. back.